compiler-core/src/io.rs

```diff
@@ -62,6 +62,62 @@ impl Utf8Writer for String {
     }
 }
 
+/// A wrapper for any writer that keeps track of what line/column it is
+/// currently on.
+///
+/// Used to produce SourceMaps.
+#[derive(Debug)]
+pub struct CursorPositionWriter<'a, W> {
+    /// The wrapper writer
+    writer: &'a mut W,
+    /// 0-indexed line the cursor is at right now
+    line: usize,
+    /// 0-indexed column the cursor is at right now
+    column: usize,
+}
+
+impl<'a, W> CursorPositionWriter<'a, W> {
+    pub fn position(&self) -> (usize, usize) {
+        (self.line, self.column)
+    }
+
+    pub fn new(writer: &'a mut W) -> Self {
+        CursorPositionWriter {
+            writer,
+            line: 0,
+            column: 0,
+        }
+    }
+}
+
+impl<'a, W: std::fmt::Write> std::fmt::Write for CursorPositionWriter<'a, W> {
+    fn write_str(&mut self, s: &str) -> std::fmt::Result {
+        self.writer.write_str(s)
+    }
+}
+
+impl<'a, W: Utf8Writer + std::fmt::Write> Utf8Writer for CursorPositionWriter<'a, W> {
+    /// A wrapper around `fmt::Write` that has Gleam's error handling.
+    fn str_write(&mut self, str: &str) -> Result<()> {
```

Comment:
This iterates the string twice. Can we do it once instead?

The pretty printer also iterates over and counts the characters, can that work be reused?

Reply:
I've reused `LineColumn` inside of this now. As long as LineNumber does not change behavior (the TODOs mention graphemes which confuses me a bit), then this should work. 

Whenever the behavior of `LineColumn` changes, the test should break :)